src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,357 @@
+
+    bool had_failure = false;
+
+    ReadOnlyBufferBuilder<ServerClusterInterface *> serverClusterBuilder;
+    ReturnErrorOnFailure(endpointProvider->ServerClusters(serverClusterBuilder));
+    auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+    for (auto * serverCluster : serverClusters)
+    {
+        VerifyOrReturnError(serverCluster != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+        CHIP_ERROR err = serverCluster->Startup(serverClusterContext);
+        if (err != CHIP_NO_ERROR)
+        {
+            had_failure = true;
+            VerifyOrDie(!serverCluster->GetPaths().empty()); // API says it must have at least one path
+
+            [[maybe_unused]] const ConcreteClusterPath path = serverCluster->GetPaths().front();
+            ChipLogError(DataManagement, "Cluster " ChipLogFormatMEI " on Endpoint %u startup failed: %" CHIP_ERROR_FORMAT,
+                         ChipLogValueMEI(path.mClusterId), path.mEndpointId, err.Format());
+        }
+    }
+    if (had_failure)
+    {
+        // If we had any failures during the startup of clusters, return an error.
+        return CHIP_ERROR_HAD_FAILURES;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ShutdownServerClusters(EndpointInterface * endpointProvider)
+{
+    VerifyOrReturnError(endpointProvider != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    ReadOnlyBufferBuilder<ServerClusterInterface *> serverClusterBuilder;
+    ReturnErrorOnFailure(endpointProvider->ServerClusters(serverClusterBuilder));
+    auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+    for (auto * serverCluster : serverClusters)
+    {
+        if (serverCluster != nullptr)
+        {
+            // TODO: The API for ServerClusterInterface->Shutdown swallows errors here. We should update it to return CHIP_ERROR.
+            serverCluster->Shutdown();
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+} // namespace
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR CodeDrivenDataModelProvider::Startup(DataModel::InteractionModelContext context)
+{
+    ReturnErrorOnFailure(DataModel::Provider::Startup(context));
+
+    mServerClusterContext.emplace(ServerClusterContext({
+        .provider           = this,
+        .storage            = mPersistentStorageDelegate,
+        .interactionContext = &mContext,
+    }));
+
+    // Startup all server clusters across all endpoints
+    bool had_failure = false;
+    for (auto * endpointProvider : mEndpointInterfaceRegistry)
+    {
+        if (StartupServerClusters(endpointProvider, *mServerClusterContext) != CHIP_NO_ERROR)
+        {
+            had_failure = true;
+        }
+    }
+
+    if (had_failure)
+    {
+        return CHIP_ERROR_HAD_FAILURES;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Shutdown()
+{
+    // Shutdown all server clusters across all endpoints
+    for (auto * endpointProvider : mEndpointInterfaceRegistry)
+    {
+        ShutdownServerClusters(endpointProvider);
```

Comment:
So this is going to call Shutdown() on cluster instances that never had Startup() called on them (if failures happen in StartupServerClusters before actually getting to the list of clusters).  This is an API contract violation; the API says:

```
    /// A shutdown will always be paired with a corresponding Startup.
```

Note that the codegen provider gets this right, as far as I can tell: it never calls Shutdown unless Startup has been called.